keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExceptionHandler.kt

```diff
@@ -58,12 +77,121 @@ class ExceptionHandler {
     return ApiError(e)
   }
 
-  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+  private fun JsonMappingException.toDetails(): ParsingErrorDetails {
+    val (rootCause, problemPath) = findRootCause()
+
+    return ParsingErrorDetails(
+      error = when (this) {
+        // caters to missing properties at the root-level (e.g. serviceAccount)
+        is MissingKotlinParameterException -> ParsingError.MISSING_PROPERTY
+        else -> when (rootCause) {
+          null -> ParsingError.OTHER
+          is MissingKotlinParameterException -> ParsingError.MISSING_PROPERTY
+          is IllegalStateException -> ParsingError.ILLEGAL_VALUE
+          is InvalidFormatException -> ParsingError.ILLEGAL_VALUE
+          is IllegalArgumentException -> ParsingError.ILLEGAL_VALUE
+          is MismatchedInputException -> ParsingError.INVALID_TYPE
+          is InvalidTypeIdException -> ParsingError.INVALID_TYPE
+          is DateTimeParseException -> ParsingError.INVALID_FORMAT
+          else -> ParsingError.OTHER
+        }
+      },
+      message = rootCause.message,
+      location = mapOf(
+        "line" to (location?.lineNr ?: -1),
+        "column" to (location?.columnNr ?: -1)
+      ),
+      path = problemPath.map { ref ->
+        val type = if (ref.from is Class<*>) {
+          ref.from as Class<*>
+        } else {
+          ref.from.javaClass
+        }
+        mapOf(
+          "type" to if (Set::class.java.isAssignableFrom(type)) {
+            // for some reason, Jackson says the type of arrays in the JSON is a java.util.HashSet
+            "array"
+          } else if (ResourceSpec::class.java.isAssignableFrom(type)) {
+            // for ResourceSpec sub-types, use the API version/kind instead of the class name
+            val handler = resourceHandlers.supporting(type as Class<ResourceSpec>)
+            if (handler != null) {
+              "${handler.supportedKind.apiVersion}/${handler.supportedKind.kind}"
+            } else {
+              type.name
+            }
+          } else {
+            type.name
+          },
+          "field" to ref.fieldName,
+          "index" to if (ref.index == -1) {
+            null
+          } else {
+            ref.index
+          }
+        )
+      }
+    )
+  }
+
+  private fun JsonMappingException.findRootCause(): Pair<Throwable, List<JsonMappingException.Reference>> {
+    if (this.cause == null) {
+      return Pair(this, this.path)
+    }
+    var exception: Throwable = this
+    var paths: MutableList<JsonMappingException.Reference> = this.path.toMutableList()
+    while (exception.cause != null && exception.cause != exception) {
+      exception = exception.cause!!
+      if (exception is JsonMappingException) {
+        paths.addAll(exception.path)
```

Comment:
This builds a flat list of all the paths leading to the leaf node/property with the problem.